test/e2e/apimachinery/watch.go

```diff
@@ -0,0 +1,237 @@
+
+		err = c.CoreV1().Pods(ns).Delete(testPodA.GetName(), nil)
+		Expect(err).NotTo(HaveOccurred())
+		expectEvent(watchA, watch.Deleted, nil)
+		expectEvent(watchAB, watch.Deleted, nil)
+		expectNoEvent(watchB, watch.Deleted, nil)
+
+		testPodB, err = c.CoreV1().Pods(ns).Create(testPodB)
+		Expect(err).NotTo(HaveOccurred())
+		expectEvent(watchB, watch.Added, testPodB)
+		expectEvent(watchAB, watch.Added, testPodB)
+		expectNoEvent(watchA, watch.Added, testPodB)
+
+		err = c.CoreV1().Pods(ns).Delete(testPodB.GetName(), nil)
+		Expect(err).NotTo(HaveOccurred())
+		expectEvent(watchB, watch.Deleted, nil)
+		expectEvent(watchAB, watch.Deleted, nil)
+		expectNoEvent(watchA, watch.Deleted, nil)
+	})
+})
+
+func watchPodsWithLabels(f *framework.Framework, labels ...string) (watch.Interface, error) {
+	c := f.ClientSet
+	ns := f.Namespace.Name
+	opts := metav1.ListOptions{
+		LabelSelector: metav1.FormatLabelSelector(&metav1.LabelSelector{
+			MatchExpressions: []metav1.LabelSelectorRequirement{
+				{
+					Key:      watchPodLabelKey,
+					Operator: metav1.LabelSelectorOpIn,
+					Values:   labels,
+				},
+			},
+		}),
+	}
+	return c.CoreV1().Pods(ns).Watch(opts)
+}
+
+func int64ptr(i int) *int64 {
+	i64 := int64(i)
+	return &i64
+}
+
+type updatePodFunc func(p *v1.Pod)
+
+func updatePod(f *framework.Framework, name string, update updatePodFunc) (*v1.Pod, error) {
+	c := f.ClientSet
+	ns := f.Namespace.Name
+	var p *v1.Pod
+	pollErr := wait.PollImmediate(2*time.Second, 1*time.Minute, func() (bool, error) {
+		var err error
+		if p, err = c.CoreV1().Pods(ns).Get(name, metav1.GetOptions{}); err != nil {
+			return false, err
+		}
+		update(p)
+		if p, err = c.CoreV1().Pods(ns).Update(p); err == nil {
+			return true, nil
+		}
+		// Only retry update on conflict
+		if !errors.IsConflict(err) {
+			return false, err
+		}
+		return false, nil
+	})
+	return p, pollErr
+}
+
+func setPodActiveDeadlineSeconds10(p *v1.Pod) {
+	p.Spec.ActiveDeadlineSeconds = int64ptr(10)
+}
+
+func setPodActiveDeadlineSeconds5(p *v1.Pod) {
+	p.Spec.ActiveDeadlineSeconds = int64ptr(5)
+}
+
+func chanThatClosesIn(seconds time.Duration) chan struct{} {
+	ch := make(chan struct{})
+	go func() {
+		time.Sleep(seconds * time.Second)
+		close(ch)
+	}()
+	return ch
+}
+
+func expectEvent(w watch.Interface, eventType watch.EventType, object runtime.Object) {
+	stop := chanThatClosesIn(10)
+	expect := watch.Event{
+		Type:   eventType,
+		Object: nil,
+	}
+	if object != nil {
+		expect.Object = object.DeepCopyObject()
+	}
+	for {
+		select {
+		case actual := <-w.ResultChan():
+			if expect.Type == actual.Type && (expect.Object == nil || apiequality.Semantic.DeepEqual(expect.Object, actual.Object)) {
+				return
+			}
+		case <-stop:
+			framework.Failf("Timed out waiting for expected event: %v", expect)
+		}
+	}
+}
+
+func expectNoEvent(w watch.Interface, eventType watch.EventType, object runtime.Object) {
+	stop := chanThatClosesIn(10)
```

Comment:
seems most code logic are the same for the two func, what about merge them into one?